src/idl_gen_cpp.cpp

```diff
@@ -2522,6 +2524,16 @@ class CppGenerator : public BaseGenerator {
                 ? ".type"
                 : (field.value.type.element == BASE_TYPE_UNION ? ".value" : "");
         code += "{ _o->" + name + ".resize(_e->size()); ";
```

Comment:
I think we want to change the `resize()` to `reserve()` in this case that we use copy. Resize initializes each element added, while reserve just allocates the memory. Now it might not make a big difference for byte[] arrays since initialization is just 0, but it might worth be testing out.

Reply:
While I agree with you, using `reserve()` actually causes the hexagon DSP to crash. In fact, if I don't `resize()`, the DSP crashes. I tested this out on godbolt.org and it *should* work fine. The Hexagon compiler and std lib implementation is... _lacking_... which is actually why I'm here making this change.

I'm really confused how reserve would not work for std::copy. We have found other oddities with the Hexagon standard library implementation as well. Another example: copying a std::vector<uint8_t> with the std::vector copy constructor is 10x slower than with std::copy or memcpy.

Reply:
It's a bit unfortunate though to use `resize` when that is not going to be required for almost every other users. A `resize` is going to touch that memory twice, and while that probably is not a big deal for small arrays it may matter for big ones.

Can we simply not call `resize` or `reserve` at all? I'd imagine `copy` does something like this internally anyway.

If not, lets just stick to using `memcpy`, if that works efficiently for all use cases.

Reply:
actually - please disregard my last comment. I'm a dufus and only just realized that the problem with `reserve()` is the code paths that skip my optimization, since they assign to the vector with `operator[]` directly.

Sorry for the confusion! I'm going to benchmark all the options and report back when I have the numbers in front of me. Might take a while, I'm trying to keep a toddler alive simultaneously.

Reply:
Initial observation: using `reserve()` as opposed to `resize()` offered no detectable latency improvement on 2MB of data (at least on my target platform). It is obviously more optimal to use `reserve()`, but I thought it worth noting that the actual copy has the dominant impact on performance.

Now I'll compare memcpy, std::copy, std::vector::insert, and std::vector::append.

Reply:
Ok, looks like std::copy is the winner here:

For 1.8MB of data on Hexagon
std::copy with std::vector::reserve: ~3.3ms
memcpy with std::vector::resize: ~3.6ms
for-loop with direct insertion and std::vector::resize: ~20ms
std::vector::insert with std::vector::reserve: ~18ms
std::vector::assign with std::vector::reserve: ~18ms

std::copy also supports more since it supports `std::vector<bool>`.

I'll update this change with the suggestions, and use the built-in `begin()` and `end()` rather than the ones in std:: to avoid problems with STLPort.

Reply:
One more observation, it looks like we actually should be using std::vector::resize before using std::copy. The alternative is to use std::back_inserter, which comes from <iterator>, but due to issues with STLPort, I think that won't work.


I'll just use resize rather than reserve, and then std::copy.

Reply:
This shouldn't be a big deal: as I mentioned previously, the performance delta between using reserve and resize was negligible for reasonably large vector copies.